### PR TITLE
Fiscal year creation with locking previous

### PR DESCRIPTION
**Pull request by mbayopanda**
This PR is enssentially for creating a new fiscal year with the previsous fiscal year closed, and with the consideration of the class 8 account.

We didn't remove the injection of $scope because whitout $scope the module doesnt work correctly

### CODE REVIEW

File: client/src/partials/fiscal/fiscal.create.js [hunk 1]
```diff
@@ -1,427 +1,446 @@
...
-        session.solde6 = session.array6.reduce(sumDebMinusCred, 0);
-        session.solde7 = session.array7.reduce(sumCredMinusDeb, 0);
-        observation();
+  // STEP 3: opening with resultat
+  function stepThree(id) {
+    var fy = getFiscalYear(id);
+    session.previous_fiscal_year = fy.fiscal_year_txt;
+    session.previous_fiscal_year_id = fy.id;
+
+    getSolde(6, fy.id)
```
Comment (DedrickEnc) [inline client/src/partials/fiscal/fiscal.create.js]: We should avoid the account_number user to keep the system independant to an accounting system

File: client/src/partials/fiscal/fiscal.create.js [hunk 2]
```diff
@@ -1,427 +1,446 @@
...
-    // fires on controller load
-    function onLoad() {
-      // filter years that are parents out of the selection
-      // in the view
-      filterParentYears();
+  // Make sure that only years without children show
+  // up in the view for selection as previous_fiscal_year
+  function filterParentYears() {
+    // copy the fiscal year store from the parent
+    var years = angular.copy(imports.fiscal.data);
```
Comment (jniles) [inline client/src/partials/fiscal/fiscal.create.js]: Do you really need `angular.copy()` here?

File: client/src/partials/fiscal/fiscal.create.js [hunk 3]
```diff
@@ -1,427 +1,446 @@
...
-        return false;
-      }
+  function closePreviousFY () {
+    var res = confirm($translate.instant('FISCAL_YEAR.CONFIRM_CLOSING'));
+    if (res) {
+      var updateFY = {
+        id : session.previous_fiscal_year_id,
+        locked : 1
+      };
+      connect.put('fiscal_year', [updateFY], ['id']);
```
Comment (jniles) [inline client/src/partials/fiscal/fiscal.create.js]: This line needs success and error handling.

File: client/src/partials/fiscal/fiscal.create.js [hunk 4]
```diff
@@ -1,427 +1,446 @@
...
 
-    function observation () {
-      if ((session.solde7 - session.solde6) > 0) {
-        session.observation = 1;
-      } else if ((session.solde7 - session.solde6) < 0)  {
-        session.observation = -1;
-      } else {
-        session.observation = 0;
-      }
+  function observation () {
```
Comment (jniles) [inline client/src/partials/fiscal/fiscal.create.js]: Could you include some comments describing what `observation()` is?

File: client/src/partials/fiscal/fiscal.create.js [hunk 5]
```diff
@@ -1,427 +1,446 @@
...
 
-    function getSolde (classe, fy) {
-      return $http.get('/getClassSolde/'+classe+'/'+fy)
-      .success(function (data) {
-        return data;
-      });
-    }
+  function getSolde (classe, fy) {
+    return $http.get('/getClassSolde/'+classe+'/'+fy)
+    .success(function (data) {
```
Comment (jniles) [inline client/src/partials/fiscal/fiscal.create.js]: Could you remove the `.success()` call here?

File: client/src/partials/fiscal/fiscal.create.js [hunk 6]
```diff
@@ -1,427 +1,446 @@
...
+    })
+    .then(function () {
+      session.solde6       = session.array6.reduce(sumDebMinusCred, 0);
+      session.solde7       = session.array7.reduce(sumCredMinusDeb, 0);
+      session.solde8Charge = session.array8Charge.reduce(sumDebMinusCred, 0);
+      session.solde8Profit = session.array8Profit.reduce(sumCredMinusDeb, 0);
+      session.charge       = session.solde6 + session.solde8Charge;
+      session.produit      = session.solde7 + session.solde8Profit;
+      observation();
+    });
```
Comment (jniles) [inline client/src/partials/fiscal/fiscal.create.js]: There is not error handling in this promise chain.  A simple `console.log(error)` is necessary and sufficient.

File: client/src/partials/fiscal/fiscal.create.js [hunk 7]
```diff
@@ -1,427 +1,446 @@
...
+        user_id   : session.user_id,
+        resultat  : {
+          resultat_account : session.resultat_account,
+          class6           : session.array6,
+          class7           : session.array7,
+          class8Charge     : session.array8Charge,
+          class8Profit     : session.array8Profit
+        }
+      };
+      $http.post('/posting_fiscal_resultat/', { params:
```
Comment (jniles) [inline client/src/partials/fiscal/fiscal.create.js]: Is there any error/success handling here?

File: client/src/partials/fiscal/fiscal.create.js [hunk 8]
```diff
@@ -1,427 +1,446 @@
...
+    $scope.step = steps[4];
+  }
 
-    // returns true if the fiscal year is for 12 months
-    function isFullYear() {
-      return data.year === 'true';
-    }
+  // returns true if the fiscal year is for 12 months
+  function isFullYear() {
+    return data.year === 'true';
```
Comment (jniles) [inline client/src/partials/fiscal/fiscal.create.js]: Why is this a string comparison?

File: client/src/partials/fiscal/fiscal.create.js [hunk 9]
```diff
@@ -1,427 +1,446 @@
...
-    // normalizes a date to a UTC date for the server
-    // TODO Put this functionality in a service
-    function normalizeUTCDate(date) {
-      var year = date.getFullYear(),
-          month = date.getMonth(),
-          day = date.getDate();
-      return Date.UTC(year, month, day);
+  // normalizes a date to a UTC date for the server
+  // TODO Put this functionality in a service
+  function normalizeUTCDate(date) {
```
Comment (jniles) [inline client/src/partials/fiscal/fiscal.create.js]: We should remove this functionality.  Just send raw dates back to the server.

**Comment by jniles** [comment]
@mbayopanda , I have made a series of comments, but they do not relate exactly to locking the previous fiscal year.  Only the comments made by @DedrickEnc and [this one](https://github.com/IMA-WorldHealth/bhima/pull/813/files#r42345382) are directly tied to locking fiscal years.  If you make a commit that addresses those, this PR can be merged.  If you prefer to fix everything, that's also excellent.  Let me know when @jniles when this is ready for another review.

**Comment by mbayopanda** [comment]
I prefer to fix it before, and you can comment then merge it after

**Comment by jniles** [comment]
Sounds good.

**Comment by jniles** [comment]
LGTM.

**Pull request merged**